### PR TITLE
JavaScript LS scaffolding + JS module inference

### PR DESCRIPTION
**Pull request by RyanCavanaugh**
This is a subset of the prior PR that carves out only the code we need to get a) JavaScript intellisense in general and b) CommonJS module inference in JavaScript files.

Includes work by @vladima in 'services' to get the correct compilation context to Roslyn

### CODE REVIEW

File: src/services/services.ts [hunk 1]
```diff
@@ -2276,41 +2227,194 @@ namespace ts {
...
+                        // record first item in the list only if its name is not "require"
+                        // record second item in the list only if its name is not "exports"
+                        // record third item in the list only if its name is not "module"
+                        // record all other items in the list unconditionally
+                        const shouldRecordName =
+                            i === 0
+                                ? moduleName !== "require"
+                                : i === 1
+                                    ? moduleName !== "exports"
+                                    : i !== 2 || moduleName !== "module";
```
Comment (DanielRosenwasser) [inline src/services/services.ts]: Consider extracting this out into a function.

```TypeScript
function shouldRecordAmdDefineName(index: number, moduleName: string) {
    // record first item in the list only if its name is not "require"
    // record second item in the list only if its name is not "exports"
    // record third item in the list only if its name is not "module"
    // record all other items in the list unconditionally
    switch (i) {
        case 0:
            return moduleName !== "require";
        case 1:
            return moduleName !== "exports";
    }
    return index !== 2 || moduleName !== "module";
}
```

Comment (RyanCavanaugh) [inline src/services/services.ts]: We can remove the AMD logic entirely

File: src/compiler/binder.ts [hunk 2]
```diff
@@ -919,6 +934,16 @@ namespace ts {
                     checkStrictModeFunctionName(<FunctionExpression>node);
                     let bindingName = (<FunctionExpression>node).name ? (<FunctionExpression>node).name.text : "__function";
                     return bindAnonymousDeclaration(<FunctionExpression>node, SymbolFlags.Function, bindingName);
+
+                case SyntaxKind.CallExpression:
+                    // We're only inspecting call expressions to detect CommonJS modules, so we can skip
+                    // this check if we've already seen the module indicator
+                    if (isJavaScriptFile && !file.commonJsModuleIndicator) {
+                        bindCallExpression(<CallExpression>node);
```
Comment (ahejlsberg) [inline src/compiler/binder.ts]: I would move the `!file.commonJsModuleIndicator` test into `bindCallExpression` so we have all the logic in one place.

File: src/compiler/checker.ts [hunk 3]
```diff
@@ -2565,6 +2572,14 @@ namespace ts {
                 if (declaration.kind === SyntaxKind.ExportAssignment) {
                     return links.type = checkExpression((<ExportAssignment>declaration).expression);
                 }
+                // Handle module.exports = expr
+                if (declaration.kind === SyntaxKind.BinaryExpression) {
+                    return links.type = checkExpression((<BinaryExpression>declaration).right);
+                }
+                // Handle exports.p = expr
+                if (declaration.kind === SyntaxKind.PropertyAccessExpression && declaration.parent.kind === SyntaxKind.BinaryExpression) {
```
Comment (ahejlsberg) [inline src/compiler/checker.ts]: Don't think you need the `declaration.parent.kind` test. There should never be a case where that isn't true.

File: src/compiler/checker.ts [hunk 4]
```diff
@@ -14881,16 +14920,24 @@ namespace ts {
...
             // Initialize global symbol table
             forEach(host.getSourceFiles(), file => {
-                if (!isExternalModule(file)) {
+                if (!isExternalOrCommonJsModule(file)) {
                     mergeSymbolTable(globals, file.locals);
                 }
             });
 
             // Initialize special symbols
+            if (compilerOptions.allowNonTsExtensions) {
```
Comment (ahejlsberg) [inline src/compiler/checker.ts]: Seems odd to have this determine whether `require` is a special global. I think it should either (a) always be a special global or (b) only be a special global in JavaScript files.

File: src/compiler/checker.ts [hunk 5]
```diff
@@ -9345,6 +9376,14 @@ namespace ts {
                     return anyType;
                 }
             }
+
+            let exprType = checkExpression(node.expression);
+            if (exprType === cjsRequireType) {
+                if (node.arguments.length === 1 && node.arguments[0].kind === SyntaxKind.StringLiteral) {
+                    return resolveExternalModuleTypeByLiteral(<StringLiteral>node.arguments[0]);
```
Comment (ahejlsberg) [inline src/compiler/checker.ts]: This happens even when the containing file is a .ts file. Is that intended?

File: src/compiler/binder.ts [hunk 6]
```diff
@@ -949,16 +973,21 @@ namespace ts {
...
+                bindSourceFileAsExternalModule();
             }
         }
 
-        function bindExportAssignment(node: ExportAssignment) {
+        function bindSourceFileAsExternalModule() {
+            bindAnonymousDeclaration(file, SymbolFlags.ValueModule, `"${removeFileExtension(file.fileName) }"`);
+        }
+
+        function bindExportAssignment(node: ExportAssignment|BinaryExpression) {
```
Comment (ahejlsberg) [inline src/compiler/binder.ts]: We usually put a blank on both sides of the `|` operator.

File: src/compiler/checker.ts [hunk 7]
```diff
@@ -160,6 +160,8 @@ namespace ts {
         let getGlobalPromiseConstructorLikeType: () => ObjectType;
         let getGlobalThenableType: () => ObjectType;
 
+        let cjsRequireType: Type;
```
Comment (ahejlsberg) [inline src/compiler/checker.ts]: We don't need this anymore, do we?

File: lib/lib.core.d.ts [hunk 8]
```diff
@@ -1225,7 +1225,7 @@ interface ArrayBuffer {
 interface ArrayBufferConstructor {
     prototype: ArrayBuffer;
     new (byteLength: number): ArrayBuffer;
-    isView(arg: any): boolean;
+    isView(arg: any): arg is ArrayBufferView;
```
Comment (ahejlsberg) [inline lib/lib.core.d.ts]: Very distracting to have all these unrelated changes.

File: src/compiler/checker.ts [hunk 9]
```diff
@@ -4176,11 +4203,15 @@ namespace ts {
...
-            let typeSymbol = namespaceSymbol && getSymbol(namespaceSymbol.exports, name, SymbolFlags.Type);
+            let typeSymbol = getExportedSymbolFromNamespace(namespace, name);
             return typeSymbol && getDeclaredTypeOfSymbol(typeSymbol);
         }
 
+        function getExportedSymbolFromNamespace(namespace: string, name: string): Symbol {
+            let namespaceSymbol = getGlobalSymbol(namespace, SymbolFlags.Namespace, /*diagnosticMessage*/ undefined);
+            return namespaceSymbol && getSymbol(namespaceSymbol.exports, name, SymbolFlags.Type | SymbolFlags.Value);
+        }
+
```
Comment (ahejlsberg) [inline src/compiler/checker.ts]: This whole change appears unnecessary. Nothing calls `getExportedSymbolFromNamespace`.

File: src/compiler/program.ts [hunk 10]
```diff
@@ -70,7 +70,7 @@ namespace ts {
     }
 
     function loadNodeModuleFromFile(candidate: string, failedLookupLocation: string[], host: ModuleResolutionHost): string {
-        return forEach(moduleFileExtensions, tryLoad);
+        return forEach(supportedJsExtensions, tryLoad);
```
Comment (ahejlsberg) [inline src/compiler/program.ts]: Is this a safe change? I don't exactly know what's going on here, but it means that we'll look for .js files as well as .ts files even in the pure TypeScript case.

File: src/compiler/program.ts [hunk 11]
```diff
@@ -724,6 +741,10 @@ namespace ts {
                         }
                         break;
                 }
+
+                if (isSourceFileJavaScript(file)) {
+                    forEachChild(node, node => collect(node, allowRelativeModuleNames));
```
Comment (ahejlsberg) [inline src/compiler/program.ts]: This isn't quite right, you'll end up looking for ES6 module constructs in nested locations where they aren't allowed. You really only should be looking for call expressions in the child nodes.

File: src/compiler/program.ts [hunk 12]
```diff
@@ -708,6 +711,20 @@ namespace ts {
                             (imports || (imports = [])).push(<LiteralExpression>moduleNameExpr);
                         }
                         break;
+                    case SyntaxKind.CallExpression:
+                        if (isJavaScriptFile && isRequireCall(node)) {
+
+                            let jsImports = (<CallExpression>node).arguments;
+                            if (jsImports) {
+                                imports = (imports || []);
+                                for (var i = 0; i < jsImports.length; i++) {
```
Comment (ahejlsberg) [inline src/compiler/program.ts]: `jsImports.length` is always one here or otherwise `isRequireCall` would have returned false. So why the loop?

File: src/compiler/utilities.ts [hunk 13]
```diff
@@ -2083,12 +2138,12 @@ namespace ts {
...
 
-    export function isJavaScript(fileName: string) {
-        return fileExtensionIs(fileName, ".js");
+    export function hasJavaScriptFileExtension(fileName: string) {
+        return fileExtensionIs(fileName, ".js") || fileExtensionIs(fileName, ".jsx");
     }
 
     export function isTsx(fileName: string) {
-        return fileExtensionIs(fileName, ".tsx");
+        return fileExtensionIs(fileName, ".tsx") || fileExtensionIs(fileName, ".jsx");
```
Comment (ahejlsberg) [inline src/compiler/utilities.ts]: Seems a little odd that something named `isTsx` return true for a .jsx extension.

File: src/harness/harness.ts [hunk 14]
```diff
@@ -211,7 +211,7 @@ namespace Utils {
             return isNodeOrArray(v) ? serializeNode(v) : v;
         }, "    ");
 
-        function getKindName(k: number | string): string {
+        function getKindName(k: number|string): string {
```
Comment (ahejlsberg) [inline src/harness/harness.ts]: Missing blanks.

File: src/compiler/checker.ts (hunk 7 above)
Comment (RyanCavanaugh) [inline src/compiler/checker.ts]: Correct; removed

File: src/compiler/checker.ts (hunk 9 above)
Comment (RyanCavanaugh) [inline src/compiler/checker.ts]: Reverted

File: lib/lib.core.d.ts (hunk 8 above)
Comment (RyanCavanaugh) [inline lib/lib.core.d.ts]: I can roll back the LKG update portion of this if it's truly problematic.

**Comment by RyanCavanaugh** [comment]
@vladima can you address the comments in program.ts?

File: src/compiler/utilities.ts (hunk 13 above)
Comment (RyanCavanaugh) [inline src/compiler/utilities.ts]: Open to suggestions on the name here. `usesJsxSyntax` ? `allowJsxExpressions` ?

File: src/compiler/program.ts (hunk 10 above)
Comment (vladima) [inline src/compiler/program.ts]: this is a good question. I do remember that we want to lift restriction for allow only `.d.ts` files and add `.ts` / `.tsx` to the mix but I'm not sure about `.js`

File: src/compiler/program.ts (hunk 12 above)
Comment (vladima) [inline src/compiler/program.ts]: done
